be/src/exec/vectorized/json_scanner.cpp

```diff
@@ -674,129 +664,359 @@ Status JsonReader::_read_and_parse_json() {
         }
 
-        column->append_strings(std::vector<Slice>{Slice(sv.data(), sv.length())});
-        break;
+        switch (type_desc.type) {
+        // Double column, float value.
+        case TYPE_DOUBLE: {
+            _construct_numeric_column(column, d);
+            return Status::OK();
+        }
+
+        // Float column, float value.
+        case TYPE_FLOAT: {
+            auto f = static_cast<float>(d);
+            _construct_numeric_column(column, f);
+            return Status::OK();
+        }
+
+        case TYPE_LARGEINT:
+        case TYPE_BIGINT:
+        case TYPE_INT:
+        case TYPE_SMALLINT:
+        case TYPE_TINYINT: {
+            std::string err_msg =
+                    strings::Substitute("Unable to construct integer column with floating value. value=$0", d);
+            return Status::DataQualityError(err_msg.c_str());
+        }
+
+        case TYPE_VARCHAR: {
+            std::ostringstream oss;
+            oss << d;
+            _construct_string_column(column, Slice{oss.str()});
+            return Status::OK();
+        }
+
+        default: {
+            std::string err_msg = strings::Substitute("Unable to cast the floating value. value=$0", d);
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        }
+
+    } else {
+        std::string_view sv;
+        simdjson::to_json_string(value).get(sv);
+        std::string err_msg =
+                strings::Substitute("unsupported value type. value=$0", std::string(sv.data(), sv.size()));
+        return Status::DataQualityError(err_msg.c_str());
     }
+}
 
-    case simdjson::ondemand::json_type::object:
-    case simdjson::ondemand::json_type::array: {
-        std::unique_ptr<char[]> buf;
-        size_t buflen{};
-        JsonFunctions::minify_json_to_string(value, buf, buflen);
+Status JsonReader::_construct_column_with_string_value(simdjson::ondemand::value& value, Column* column,
+                                                       const TypeDescriptor& type_desc) {
+    std::string_view sv;
+    auto err = value.get_string().get(sv);
+    if (UNLIKELY(err)) {
+        std::string err_msg = strings::Substitute("Failed to parse the string column value. code=$0, error=$1", err,
+                                                  simdjson::error_message(err));
+        return Status::DataQualityError(err_msg.c_str());
+    }
 
-        column->append_strings(std::vector<Slice>{Slice{buf.get(), buflen}});
-        break;
+    switch (type_desc.type) {
+    case TYPE_VARCHAR: {
+        _construct_string_column(column, Slice{sv.data(), sv.size()});
+        return Status::OK();
+    }
+
+    case TYPE_LARGEINT: {
+        StringParser::ParseResult parse_result = StringParser::PARSE_SUCCESS;
+        auto i128 = StringParser::string_to_int<int128_t>(sv.data(), sv.length(), &parse_result);
+        if (parse_result == StringParser::PARSE_SUCCESS) {
+            _construct_numeric_column(column, i128);
+        } else {
```

Comment:
It's possible that `sv` is a floating-point value, in this case, the `StringParser::string_to_int` will fail, but the original implementation based on cast expression will succeed.

You can take the CSV code as a reference:
https://github.com/StarRocks/starrocks/blob/31916cbdcf78df6945d200ab71533b889ff0de31/be/src/formats/csv/numeric_converter.cpp#L35
https://github.com/StarRocks/starrocks/blob/31916cbdcf78df6945d200ab71533b889ff0de31/be/test/formats/csv/numeric_converter_test.cpp#L61